cpp/ql/src/semmle/code/cpp/ir/internal/CppType.qll

```diff
@@ -0,0 +1,500 @@
+    type instanceof UnknownType and isGLValue = true
+  }
+}
+
+/**
+ * A `CppType` that represents an unknown type.
+ */
+private class CppUnknownType extends CppType, TUnknownType {
+  override final string toString() {
+    result = any(UnknownType type).toString()
+  }
+
+  override final IRUnknownType getIRType() {
+    any()
+  }
+
+  override final predicate hasType(Type type, boolean isGLValue) {
+    type instanceof UnknownType and isGLValue = false
+  }
+}
+
+/**
+ * Gets the single instance of `CppUnknownType`.
+ */
+CppUnknownType getUnknownType() {
+  any()
+}
+
+/**
+ * Gets the `CppType` that represents a prvalue of type `void`.
+ */
+CppPRValueType getVoidType() {
+  exists(VoidType voidType |
+    result.hasType(voidType, false)
+  )
+}
+
+/**
+ * Gets the `CppType` that represents a prvalue of type `type`.
+ */
+CppType getTypeForPRValue(Type type) {
+  if type instanceof UnknownType
+  then result instanceof CppUnknownType
+  else result.hasType(type, false)
+}
+
+/**
+ * Gets the `CppType` that represents a glvalue of type `type`.
+ */
+CppType getTypeForGLValue(Type type) {
+  result.hasType(type, true)
+}
+
+/**
+ * Gets the `CppType` that represents a prvalue of type `int`.
+ */
+CppPRValueType getIntType() {
+  exists(IntType type |
+    type.isImplicitlySigned() and
+    result.hasType(type, false)
+  )
+}
+
+/**
+ * Gets the `CppType` that represents a prvalue of type `bool`.
+ */
+CppPRValueType getBoolType() {
+  exists(BoolType type |
+    result.hasType(type, false)
+  )
+}
+
+/**
+ * Gets the `CppType` that represents a glvalue of function type.
+ */
+CppFunctionGLValueType getFunctionGLValueType() {
+  any()
+}
+
+/**
+ * Gets the `CppType` that represents a blob of unknown type with size `byteSize`.
+ */
+CppUnknownBlobType getUnknownBlobType(int byteSize) {
+  result.getByteSize() = byteSize
+}
+
+/**
+ * Gets the `CppType` that is the canonical type for an `IRBooleanType` with the specified
+ * `byteSize`.
+ */
+CppWrappedType getCanonicalBooleanType(int byteSize) {
+  exists(BoolType type |
+    result = TPRValueType(type) and byteSize = type.getSize()
+  )
+}
+
+/**
+ * Compute the sorting priority of an `IntegralType` based on its signedness.
+ */
+private int getSignPriority(IntegralType type) {
+  /*
+    Explicitly unsigned types sort first. Explicitly signed types sort last. Types with no explicit
+    signedness sort in between. This lets us always choose `int` over `signed int`, while also
+    choosing `unsigned char`+`char` when `char` is signed, and `unsigned char`+`signed char` when
+    `char` is unsigned.
+  */
+  if type.isExplicitlyUnsigned()
+  then result = 2
+  else if type.isExplicitlySigned()
+  then result = 0
+  else result = 1
+}
+
+/**
+ * Gets the sort priority of an `IntegralType` based on its kind.
+ */
+private int getKindPriority(IntegralType type) {
+  /*
+    `CharType` sorts lower so that we prefer the plain integer types when they have the same size as
+    a `CharType`.
+  */
```

Comment:
These comments will not autoformat well. Comments in this location should be of the `//` type, or the autoformatter will put a blank line after them. Same above.

Reply:
Fixed

Reply:
It's not fixed. There is too much whitespace within and after the comment.

Reply:
Weird. I must have changed a different comment. Fix for real this time.